lib/Collaboration/Reference/TalkReferenceProvider.php

```diff
@@ -0,0 +1,288 @@
+			$comment = $this->chatManager->getComment($room, $referenceMatch['message']);
+			$message = $this->messageParser->createMessage($room, $participant, $comment, $this->l);
+			$this->messageParser->parseMessage($message);
+
+			$placeholders = $replacements = [];
+			foreach ($message->getMessageParameters() as $placeholder => $parameter) {
+				$placeholders[] = '{' . $placeholder . '}';
+				if ($parameter['type'] === 'user' || $parameter['type'] === 'guest') {
+					$replacements[] = '@' . $parameter['name'];
+				} else {
+					$replacements[] = $parameter['name'];
+				}
+			}
+			$description = str_replace($placeholders, $replacements, $message->getMessage());
+
+			$titleLine = $this->l->t('Message of {user} in {conversation}');
+			if ($room->getType() === Room::TYPE_ONE_TO_ONE) {
+				$titleLine = $this->l->t('Message of {user}');
+			}
+
+			$displayName = $message->getActorDisplayName();
+			if ($message->getActorType() === Attendee::ACTOR_GUESTS) {
+				if ($displayName === '') {
+					$displayName = $this->l->t('Guest');
+				} else {
+					$displayName = $this->l->t('%s (guest)', $displayName);
+				}
+			} elseif ($displayName === '') {
+				$titleLine = $this->l->t('Message of a deleted user in {conversation}');
+			}
+
+			$title = str_replace(
+				['{user}', '{conversation}'],
+				[$displayName, $title],
+				$titleLine
+			);
+		}
+
+		$reference->setTitle($title);
+		$reference->setDescription($description);
+		$reference->setUrl($this->urlGenerator->linkToRouteAbsolute('spreed.Page.showCall', ['token' => $room->getToken()]));
+		$reference->setImageUrl($this->getRoomIconUrl($room, $this->userId));
+
+		$reference->setRichObject('call', [
+			'id' => $room->getToken(),
+			'name' => $roomName,
+			'link' => $reference->getUrl(),
+			'call-type' => $this->getRoomType($room),
+		]);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getCachePrefix(string $referenceId): string {
+		$referenceMatch = $this->getTalkAppLinkToken($referenceId);
+		if ($referenceMatch === null) {
+			return '';
+		}
+
+		return $referenceMatch['token'];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getCacheKey(string $referenceId): ?string {
+		$referenceMatch = $this->getTalkAppLinkToken($referenceId);
+		if ($referenceMatch === null) {
+			return '';
+		}
+
+		return ($this->userId ?? '') . '#' . ($referenceMatch['message'] ?? 0);
+	}
+
+	protected function getRoomIconUrl(Room $room, string $userId): string {
+		if ($room->getType() === Room::TYPE_ONE_TO_ONE) {
+			$participants = json_decode($room->getName(), true);
+
+			foreach ($participants as $p) {
+				if ($p !== $userId) {
+					return $this->urlGenerator->linkToRouteAbsolute(
+						'core.avatar.getAvatar',
+						[
+							'userId' => $p,
+							'size' => 64,
+						]
+					);
+				}
+			}
+		}
+
+		return $this->urlGenerator->getAbsoluteURL($this->urlGenerator->imagePath('spreed', 'changelog.svg'));
+	}
+
+	protected function getRoomType(Room $room): string {
```

Comment:
We already have other method very similar. Isn't the moment to move this to a more generic place and cover with unit test?
https://github.com/nextcloud/spreed/blob/758a5a767e61e2504152d86b7d051f89396ff882/lib/Notification/Notifier.php#L548-L559

Reply:
The function is now only chained 3rdparty calls, so not sure tests are too useful.